sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala

```diff
@@ -310,13 +310,19 @@ case class StructType(fields: Array[StructField]) extends DataType with Seq[Stru
   protected[sql] def toAttributes: Seq[AttributeReference] =
     map(f => AttributeReference(f.name, f.dataType, f.nullable, f.metadata)())
 
-  def treeString: String = {
+  def treeString: String = treeString(Int.MaxValue)
+
+  def treeString(level: Int): String = {
     val builder = new StringBuilder
     builder.append("root\n")
     val prefix = " |"
     fields.foreach(field => field.buildFormattedString(prefix, builder))
 
-    builder.toString()
+    if (level <= 0 || level == Int.MaxValue) {
+      builder.toString()
+    } else {
+      builder.toString().split("\n").filter(_.lastIndexOf("|--") < level * 5 + 1).mkString("\n")
```

Comment:
Sorry if I am late to the party. This seems incredibly brittle, can avoid creating the string in the first place?

Reply:
Do you mean avoid `builder.toString()` part in this line?

Reply:
Or, reconstructing line 314 ~ 317?

Reply:
The latter. I know that requires us to pass the depth around and can be a bit annoying, but I do feel that is more robust than a post-hoc step.

Reply:
Got it. Let me think about that way.